### PR TITLE
Added current grid icon a grigs page, added rules management, changed rules

### PR DESCRIPTION
**Pull request by Hukaparya**
Added rules management and changed rules:

image

Added current grid on choose grids page

image

### CODE REVIEW

File: lib/grids_page.dart [hunk 1]
```diff
@@ -82,8 +87,10 @@ class SelectGrid extends StatelessWidget {
       cursor: SystemMouseCursors.click,
       child: GestureDetector(
         onTap: () {
-          currentGrid.value = grid;
-          Navigator.pop(context);
+          if (grid.length > boardSize) {
```
Comment (polina-c) [inline lib/grids_page.dart:90]: Can you add comment why we need this condition?

**Comment by polina-c** [comment]
Let's remove identical grid (misere tic-tac-toe).
Also let's remove rule from grid name. Rename 'Four in a row' to '7x6'

File: lib/grids_page.dart (hunk 1 above)
Comment (Hukaparya) [reply lib/grids_page.dart:90]: Sorry, we don't need it, I thought to make changing of avaliable in-a-row depending of the board size, but made it easier way

**Comment by polina-c** [comment]
Please, remove existing rules explanation completely.
And, inline controls with explanation:
'Get <dropbox with number> in a row to <dropbox with two items: win, lose>.

**Comment by Hukaparya** [comment]
Changed after review

image

**Comment by polina-c** [comment]
Please, change 'In a row' to 'in a row'

File: lib/model/boards.dart [hunk 2]
```diff
@@ -131,16 +133,15 @@ final boardTemplates = <String, Board>{
...
-    [CellState.empty, CellState.empty, CellState.empty],
-    [CellState.empty, CellState.empty, CellState.empty],
-  ],
 };
 
 /// Creates a new board from an existing board template named [boardName].
 Board createBoard(String boardName) {
+  clickable = true;
+  winner = '';
+
```
Comment (polina-c) [inline lib/model/boards.dart:142]: Why we need these variables?

File: lib/model/boards.dart [hunk 3]
```diff
@@ -131,16 +133,15 @@ final boardTemplates = <String, Board>{
...
   final template = boardTemplates[boardName]!;
+  boardSize = template.length;
   return List.generate(
```
Comment (polina-c) [inline lib/model/boards.dart:145]: Why we need this variable?

File: lib/model/game.dart [hunk 4]
```diff
@@ -29,6 +34,7 @@ class Game {
     board[x][y] = result;
 
     lastMove.value = (x: x, y: y);
+    // winner = detectWinner(x, y, board);
```
Comment (polina-c) [inline lib/model/game.dart:37]: Let's delete this comment

File: lib/widgets/game_view.dart [hunk 5]
```diff
@@ -1,11 +1,28 @@
...
+int whoWins = 0;
+
+bool _clickable = true;
+
+bool get clickable => _clickable;
+
+set clickable(bool value) {
+  _clickable = value;
+}
+
```
Comment (polina-c) [inline lib/widgets/game_view.dart:25]: Most of the added items are not needed. Remove them.
Please, add doc comments to needed ones to explain what they mean and how they are used.

File: lib/widgets/game_view.dart [hunk 6]
```diff
@@ -84,6 +101,42 @@ class _GameViewState extends State<GameView> {
           },
           child: const Text('Choose grid'),
         ),
+        const SizedBox(height: 20),
+        Visibility(
+            visible: winner != '',
```
Comment (polina-c) [inline lib/widgets/game_view.dart:106]: We are not detecting winner in this PR. Please remove the code related to it.

File: lib/widgets/rules_buttons.dart [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+  Widget build(BuildContext context) {
+    return Row(
+      mainAxisAlignment: MainAxisAlignment.center,
+      children: [
+        const Text('Get'),
+        const SizedBox(width: 20),
+        SizedBox(
+          width: 87,
+          height: 30, // Replace with the desired width
+          child: OutlinedButton(
```
Comment (polina-c) [inline lib/widgets/rules_buttons.dart:25]: Why we use button to wrap another button?

File: lib/model/game.dart [hunk 8]
```diff
@@ -1,4 +1,6 @@
 import 'package:flutter/material.dart';
+// import 'package:tic_tac_toe/model/detect_winner.dart';
```
Comment (polina-c) [inline lib/model/game.dart:2]: Remove comment

File: lib/model/rules.dart
Comment (polina-c) [inline lib/model/rules.dart:1]: We do not need this file any more. Let's delete it.

File: lib/model/boards.dart (hunk 2 above)
Comment (Hukaparya) [reply lib/model/boards.dart:142]: removed

File: lib/model/boards.dart (hunk 3 above)
Comment (Hukaparya) [reply lib/model/boards.dart:145]: removed

File: lib/widgets/game_view.dart (hunk 5 above)
Comment (Hukaparya) [reply lib/widgets/game_view.dart:25]: removed

Comment (Hukaparya) [reply lib/widgets/game_view.dart:25]: removed

File: lib/widgets/game_view.dart (hunk 6 above)
Comment (Hukaparya) [reply lib/widgets/game_view.dart:106]: removed

File: lib/widgets/rules_buttons.dart (hunk 7 above)
Comment (Hukaparya) [reply lib/widgets/rules_buttons.dart:25]: easier to change size

Comment (Hukaparya) [reply lib/widgets/rules_buttons.dart:25]: flutter says that better to replace Container with SizedBox

Comment (polina-c) [reply lib/widgets/rules_buttons.dart:25]: I still see OutlinedButton in the code.
Container can do both size and border.
Please, do it for both drop downs.

File: lib/widgets/rules_buttons.dart [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+                  child: Text('3'),
+                ),
+                if (boardSize > 3)
+                  const DropdownMenuItem<int>(
+                    value: 4,
+                    child: Text('4'),
+                  ),
+                if (boardSize > 4)
+                  const DropdownMenuItem<int>(
+                    value: 5,
```
Comment (polina-c) [inline lib/widgets/rules_buttons.dart:52]: Please, make values to be 2 to 10 and add them in loop.

File: lib/widgets/game_view.dart [hunk 10]
```diff
@@ -1,11 +1,20 @@
...
+import 'package:tic_tac_toe/widgets/rules_buttons.dart';
 
 import '../grids_page.dart';
 import '../model/boards.dart';
 import '../model/game.dart';
 import '../styles/colors.dart';
 import 'board.dart';
 
+var winner = '';
+
```
Comment (polina-c) [inline lib/widgets/game_view.dart:11]: Can you review code and remove everything that relates to winner detection?

File: lib/widgets/board_icon.dart [hunk 11]
```diff
@@ -8,10 +9,13 @@ import '../model/game.dart';
 /// It holds board value named [board].
 class BoardIconView extends StatelessWidget {
   const BoardIconView({super.key, required this.board});
+
   final Board board;
 
   @override
   Widget build(BuildContext context) {
+    final int boardLength = board.length;
+    boardSize = boardLength;
```
Comment (polina-c) [inline lib/widgets/board_icon.dart:18]: Do not store boardSize as global variable. Pass it to constructor of RulesChangingButtons instead.

File: lib/widgets/board_icon.dart [hunk 12]
```diff
@@ -8,10 +9,13 @@ import '../model/game.dart';
 /// It holds board value named [board].
 class BoardIconView extends StatelessWidget {
   const BoardIconView({super.key, required this.board});
+
   final Board board;
 
   @override
   Widget build(BuildContext context) {
+    final int boardLength = board.length;
```
Comment (polina-c) [inline lib/widgets/board_icon.dart:17]: Why length?
What is possible is hard to detect programmatically, because board may contain holes. 
So let's for now just allow to select 2..10 in-a-row. And it is ok if it is not possible.

File: lib/widgets/game_view.dart [hunk 13]
```diff
@@ -1,11 +1,20 @@
...
+late final Game game1;
+int inRowValue = 3;
```
Comment (polina-c) [inline lib/widgets/game_view.dart:13]: These two variables should not be global variables, but part of state of the widget RulesChangingButtons.

File: lib/widgets/game_view.dart [hunk 14]
```diff
@@ -1,11 +1,20 @@
...
 
 import '../grids_page.dart';
 import '../model/boards.dart';
 import '../model/game.dart';
 import '../styles/colors.dart';
 import 'board.dart';
 
+var winner = '';
+
+late final Game game1;
```
Comment (polina-c) [inline lib/widgets/game_view.dart:12]: Is it needed?

Comment (Hukaparya) [reply lib/widgets/game_view.dart:10]: removed

File: lib/widgets/game_view.dart (hunk 13 above)
Comment (Hukaparya) [reply lib/widgets/game_view.dart:11]: we dont need it if we are not checking the board size. removed

File: lib/widgets/board_icon.dart (hunk 12 above)
Comment (Hukaparya) [reply lib/widgets/board_icon.dart]: done

File: lib/grids_page.dart [hunk 15]
```diff
@@ -57,6 +57,10 @@ class GridsPage extends StatelessWidget {
                     },
                     child: const Text('Cancel'),
                   ),
+                  const SizedBox(height: 20),
+                  const Text('Current Grid:'),
+                  const SizedBox(height: 20),
+                  SelectGrid(grid: grid.value, currentGrid: grid)
                 ]),
```
Comment (polina-c) [inline lib/grids_page.dart:64]: Instead of duplicating current grid, let's just highlight is with different background.
Let's do it in different PR, to simplify this PR and merge it earlier.

File: lib/widgets/game_view.dart (hunk 13 above)
Comment (polina-c) [reply lib/widgets/game_view.dart:10]: I still see them in PR.

File: lib/widgets/rules_buttons.dart [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+
+class RulesChangingButtonsState extends State<RulesChangingButtons> {
+  @override
+  Widget build(BuildContext context) {
+    return Row(
+      mainAxisAlignment: MainAxisAlignment.center,
+      children: [
+        const Text('Get'),
+        const SizedBox(width: 20),
+        SizedBox(
```
Comment (polina-c) [inline lib/widgets/rules_buttons.dart:20]: This is a big code duplicate.
Can you create a widget like RoundedDropdown, that will take items and onChanged as parameters, and use it two times?

Comment (Hukaparya) [reply lib/widgets/rules_buttons.dart:24]: Done

File: lib/main.dart [hunk 17]
```diff
@@ -54,7 +53,6 @@ class _MyHomePageState extends State<MyHomePage> {
                 children: <Widget>[
                   GameView(),
                   SizedBox(height: 40),
```
[the remark this answers is not in the document]
Comment (Hukaparya) [reply lib/main.dart]: Done

File: lib/grids_page.dart (hunk 1 above)
Comment (Hukaparya) [reply lib/grids_page.dart]: Deleted.

File: lib/widgets/rules_buttons.dart (hunk 7 above)
Comment (Hukaparya) [reply lib/widgets/rules_buttons.dart]: Now its made another way

File: lib/model/game.dart (hunk 8 above)
Comment (Hukaparya) [reply lib/model/game.dart]: Done

File: lib/model/rules.dart
Comment (Hukaparya) [reply lib/model/rules.dart:1]: Deleted

File: lib/widgets/rules_buttons.dart (hunk 9 above)
Comment (Hukaparya) [reply lib/widgets/rules_buttons.dart]: Now it's made other way

File: lib/widgets/game_view.dart (hunk 10 above)
Comment (Hukaparya) [reply lib/widgets/game_view.dart]: Done

File: lib/widgets/game_view.dart (hunk 13 above)
Comment (Hukaparya) [reply lib/widgets/game_view.dart:10]: Deleted

File: lib/grids_page.dart (hunk 15 above)
Comment (Hukaparya) [reply lib/grids_page.dart:64]: Done in another PR

File: lib/widgets/game_view.dart (hunk 13 above)
Comment (polina-c) [reply lib/widgets/game_view.dart:10]: I still see global variables in PR:

```
int inRowValue = 3;
```

All them should be moved to state classes of widgets, if they are needed.

**Comment by polina-c** [comment]
Please, resolve conflicts: 
Screenshot 2023-09-07 at 1 02 04 PM

File: lib/widgets/game_view.dart (hunk 13 above)
Comment (Hukaparya) [reply lib/widgets/game_view.dart]: Moved it to the widget

File: lib/widgets/game_view.dart [hunk 18]
```diff
@@ -7,12 +7,8 @@ import '../model/game.dart';
 import '../styles/colors.dart';
 import 'board.dart';
 
-var winner = '';
-
-late final Game game1;
-int inRowValue = 3;
 bool misere = false;
```
Comment (polina-c) [inline lib/widgets/game_view.dart:10]: These globals should go away too:

```
bool misere = false;

int whoWins = 0;
```

File: lib/widgets/rules_buttons.dart [hunk 19]
```diff
@@ -1,7 +1,12 @@
 import 'package:flutter/material.dart';
+import 'package:tic_tac_toe/widgets/dropdown.dart';
 
 import 'game_view.dart';
 
+String selectedValue = '3';
```
Comment (polina-c) [inline lib/widgets/rules_buttons.dart:6]: These globals should go away too.

There should not be global variables at all in the application.
(Global constants ar ok)

**Review by polina-c — CHANGES_REQUESTED** [review]
All global variables should go away. Let me know if you do not know how to do it.

File: lib/widgets/game_view.dart (hunk 18 above)
Comment (Hukaparya) [reply lib/widgets/game_view.dart]: Moved vars to Widget class

File: lib/widgets/rules_buttons.dart (hunk 19 above)
Comment (Hukaparya) [reply lib/widgets/rules_buttons.dart]: Moved vars to Widget class

File: lib/widgets/rules_buttons.dart [hunk 20]
```diff
@@ -39,6 +43,8 @@ class RulesChangingButtonsState extends State<RulesChangingButtons> {
                   setState(() {
                     selectedValue = value ?? '3';
                     inRowValue = int.tryParse(value ?? '0') ?? 0;
+                    // final whoWins = value == 'Win' ? 0 : 1;
```
Comment (polina-c) [inline lib/widgets/rules_buttons.dart:46]: Remove the comment.

File: lib/widgets/rules_buttons.dart [hunk 21]
```diff
@@ -64,7 +70,8 @@ class RulesChangingButtonsState extends State<RulesChangingButtons> {
                 onChanged: (value) {
                   setState(() {
                     winValue = value ?? 'Win';
```
Comment (polina-c) [inline lib/widgets/rules_buttons.dart:72]: Make 'Win' and 'Lose' constants, so that mistake in spelling would be noticed by compiler.

File: lib/widgets/rules_buttons.dart [hunk 22]
```diff
@@ -1,20 +1,24 @@
...
 
   @override
   RulesChangingButtonsState createState() => RulesChangingButtonsState();
 }
 
 class RulesChangingButtonsState extends State<RulesChangingButtons> {
+  String selectedValue = '3';
+  String winValue = 'Win';
+  int inRowValue = 3;
+  int whoWins = 0;
```
Comment (polina-c) [inline lib/widgets/rules_buttons.dart:20]: Replace whoWins to boolean with name 'rowBuilderWins', to make it more understandable.

File: lib/widgets/game_view.dart [hunk 23]
```diff
@@ -90,7 +88,14 @@ class _GameViewState extends State<GameView> {
           child: const Text('Choose grid'),
         ),
         const SizedBox(height: 20),
-        const RulesChangingButtons(),
+        RulesChangingButtons(
+          onRulesChanged: (inRowValue, whoWins) {
+            setState(() {
+              inRowValu = inRowValue;
```
Comment (polina-c) [inline lib/widgets/game_view.dart:94]: rename to `inRowValue` and use this.inRowValue to avoid name collision.

**Review by polina-c — CHANGES_REQUESTED** [review]

File: lib/widgets/rules_buttons.dart (hunk 20 above)
Comment (Hukaparya) [reply lib/widgets/rules_buttons.dart:46]: Done

File: lib/widgets/rules_buttons.dart (hunk 22 above)
Comment (Hukaparya) [reply lib/widgets/rules_buttons.dart:20]: Done

File: lib/widgets/game_view.dart (hunk 23 above)
Comment (Hukaparya) [reply lib/widgets/game_view.dart:94]: Done

File: lib/widgets/rules_buttons.dart (hunk 21 above)
Comment (Hukaparya) [reply lib/widgets/rules_buttons.dart:70]: items: const ['Win', 'Lose']

File: lib/widgets/game_view.dart [hunk 24]
```diff
@@ -91,8 +91,8 @@ class _GameViewState extends State<GameView> {
         RulesChangingButtons(
           onRulesChanged: (inRowValue, whoWins) {
             setState(() {
-              inRowValu = inRowValue;
-              whoWin = whoWins;
+              this.inRowValue = inRowValue;
+              rowBuilderWins = whoWins;
```
Comment (polina-c) [inline lib/widgets/game_view.dart:95]: Please rename parameter to rowBuilderWins too.

File: lib/widgets/game_view.dart [hunk 25]
```diff
@@ -15,8 +15,8 @@ class GameView extends StatefulWidget {
 }
 
 class _GameViewState extends State<GameView> {
-  int whoWin = 0;
-  int inRowValu = 3;
+  bool rowBuilderWins = false;
```
Comment (polina-c) [inline lib/widgets/game_view.dart:18]: Let it default to true, a traditional game.

File: lib/widgets/rules_buttons.dart [hunk 26]
```diff
@@ -43,8 +42,7 @@ class RulesChangingButtonsState extends State<RulesChangingButtons> {
                   setState(() {
                     selectedValue = value ?? '3';
                     inRowValue = int.tryParse(value ?? '0') ?? 0;
```
Comment (polina-c) [inline lib/widgets/rules_buttons.dart:44]: Why value is string? Can we make it int, so that we do not need to parse?

File: lib/widgets/rules_buttons.dart (hunk 21 above)
Comment (polina-c) [reply lib/widgets/rules_buttons.dart:70]: I mean different thing. I see 'Win' is repeated in the code. This is red flag, because if you change it in one place to for example 'Win!', and forget to change in others, the code will stop working.

And in this case it is better to have value boolean, not string. 'Win' should be just text to display.

**Review by polina-c — CHANGES_REQUESTED** [review]

File: lib/widgets/rules_buttons.dart (hunk 26 above)
Comment (Hukaparya) [reply lib/widgets/rules_buttons.dart:44]: We have universal dropdown widget, that takes strings as an argument and returns string. And then we change this string to type we want (in this case int, in another bool).

File: lib/widgets/game_view.dart (hunk 25 above)
Comment (Hukaparya) [reply lib/widgets/game_view.dart:18]: Done

File: lib/widgets/rules_buttons.dart [hunk 27]
```diff
@@ -0,0 +1,85 @@
...
+    required this.onRulesChanged,
+  }) : super(key: key);
+
+  @override
+  RulesChangingButtonsState createState() => RulesChangingButtonsState();
+}
+
+class RulesChangingButtonsState extends State<RulesChangingButtons> {
+  String selectedValue = '3';
+  String winValue = 'Win';
```
Comment (polina-c) [inline lib/widgets/rules_buttons.dart:18]: Use constant here.

File: lib/widgets/rules_buttons.dart [hunk 28]
```diff
@@ -0,0 +1,85 @@
...
+    Key? key,
+    required this.onRulesChanged,
+  }) : super(key: key);
+
+  @override
+  RulesChangingButtonsState createState() => RulesChangingButtonsState();
+}
+
+class RulesChangingButtonsState extends State<RulesChangingButtons> {
+  String selectedValue = '3';
```
[the remark this answers is not in the document]
Comment (Hukaparya) [reply lib/widgets/rules_buttons.dart:17]: Done

File: lib/widgets/rules_buttons.dart (hunk 27 above)
Comment (Hukaparya) [reply lib/widgets/rules_buttons.dart:18]: Done

**Review by polina-c — APPROVED** [review]
